libretro-common/playlists/label_sanitization.c

```diff
@@ -0,0 +1,168 @@
+/* Copyright  (C) 2010-2019 The RetroArch team
+ *
+ * ---------------------------------------------------------------------------------------
+ * The following license statement only applies to this file (file_path.h).
+ * ---------------------------------------------------------------------------------------
+ *
+ * Permission is hereby granted, free of charge,
+ * to any person obtaining a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <playlists/label_sanitization.h>
+#include <retro_miscellaneous.h>
+#include <string/stdstring.h>
+#include <string.h>
+
+/*
+ * Does not work with nested blocks.
+ */
+void label_sanitize(char *label, size_t size, bool (*left)(char*), bool (*right)(char*))
+{
+   bool copy = true;
+   int rindex = 0;
+   int lindex = 0;
+   char new_label[PATH_MAX_LENGTH];
+
+   for (; lindex < size && label[lindex] != '\0'; lindex++)
+   {
+      if (copy)
+      {
+         /* check for the start of the range */
+         if ((*left)(&label[lindex]))
+            copy = false;
+
+         if (copy)
+            new_label[rindex++] = label[lindex];
+      }
+      else if ((*right)(&label[lindex]))
+         copy = true;
+   }
+
+   new_label[rindex] = label[lindex];
+
+   strcpy(label, new_label);
```

Comment:
Should probably be a little bit more careful here :)

- If input `label` is not null-terminated, resultant `new_label` won't be null-terminated either.

- If input `size` is greater than `PATH_MAX_LENGTH` it's possible to generate a segfault (unlikely, but possible)

- Should always use `strncpy()` rather than `strcpy()`